salt/roster/cache.py

```diff
@@ -1,60 +1,217 @@
+This roster module can query the Salt Master's cache to access minions over ``salt-ssh``.
 
-Currently only contains logic to return an IPv4 address; does not handle IPv6,
-or authentication (passwords, keys, etc).
+The roster uses the configuration option ``roster_order`` to get a list of hosts from the minion data.
 
-It is possible to configure this roster to prefer a particular type of IP over
-another. To configure the order, set the roster_order in the master config
-file. The default for this is:
+All matched minions will have the SSH options from ``roster_defaults`` in the master configuration, appended with
+the config options found in ``roster_order``.
 
+The default configuration is:
 .. code-block:: yaml
+roster_order:
+    host:
+      - ipv6-private  # IPv6 addresses in private ranges
+      - ipv6-global   # IPv6 addresses in global ranges
+      - ipv4-private  # IPv4 addresses in private ranges
+      - ipv4-public   # IPv4 addresses in public ranges
+      - ipv4-local    # loopback addresses
+
+
+Various other keys are also supported:
+.. code-block:: yaml
+
+# Examples
+roster_order:
+  host:
+    - id                             # use the minion id
+    - global|public|private|local    # Both IPv6 and IPv4 addresses in that range
+    - grain: fqdn_ip4                # lookup this grain
+    - pillar:                        # Works with pillardata too
+        - dns:fqdn                   # ':' syntax and lists supported
+    - mine: network.ip_addrs         # Mine data lookup works the same
+    - 2000::/3                       # CIDR networks, both IPv4 and IPv6 supported
+
+                                     # The lookup can be used for arbitrary other keys as well,
+  user: pillar: ssh:auth:user
+  priv:
+    - pillar: ssh:auth:private_key
+
+
+.. versionadded:: Nitrogen
+    expanded functionality to include arbitrary pillar and grain keys, IPv6 addresses
+    allow roster_order per key in config
+    changed default order in line with os defaults
+
 
-    roster_order:
-      - public
-      - private
-      - local
 '''
 from __future__ import absolute_import
 
-# Import Salt libs
-import salt.loader
-import salt.utils
-import salt.utils.cloud
-import salt.utils.validate.net
+# Python
+import re
+
+# Salt libs
+import salt.utils.minions
+import salt.cache
+from salt._compat import ipaddress
+import salt.ext.six as six
+
+# Debug & Logging
+import logging
+#import pprint
+
+log = logging.getLogger(__name__)
+#ppr = pprint.PrettyPrinter(indent=2).pprint
 
 
 def targets(tgt, tgt_type='glob', **kwargs):  # pylint: disable=W0613
     '''
-    Return the targets from the flat yaml file, checks opts for location but
-    defaults to /etc/salt/roster
+    Return the targets from the minion cache
+    evaluate roster_order when determining config values
     '''
+    minions = salt.utils.minions.CkMinions(__opts__)
+    minions = minions.check_minions(tgt, tgt_type)
```

Comment:
I went digging through `utils` and dug this one up, didn't know specifically about the runner ;) 

Since it's a `cache` roster I never really looked at using live data and it turns out the `cache` runner does exactly this, only through the `MasterPillarUtil` util.

So no particular purpose, and I've no qualm about refactoring it though if you guys insist but also fail to see the point ;)